pkg/util/loop/loop.go

```diff
@@ -160,37 +179,36 @@ func (loop *Device) AttachFromFile(image *os.File, mode int, number *int) error
 				return nil
 			}
 			syscall.Close(loopFd)
+			continue
 		} else {
 			_, _, esys := syscall.Syscall(syscall.SYS_IOCTL, uintptr(loopFd), CmdSetFd, image.Fd())
 			if esys != 0 {
 				syscall.Close(loopFd)
 				continue
 			}
-			break
 		}
-	}
 
-	if _, _, err := syscall.Syscall(syscall.SYS_FCNTL, uintptr(loopFd), syscall.F_SETFD, syscall.FD_CLOEXEC); err != 0 {
-		return fmt.Errorf("failed to set close-on-exec on loop device %s: %s", path, err.Error())
-	}
+		if _, _, err := syscall.Syscall(syscall.SYS_FCNTL, uintptr(loopFd), syscall.F_SETFD, syscall.FD_CLOEXEC); err != 0 {
+			return fmt.Errorf("failed to set close-on-exec on loop device %s: %s", path, err.Error())
+		}
 
-	maxRetries := 5
-	for i := 0; i < maxRetries; i++ {
 		if _, _, err := syscall.Syscall(syscall.SYS_IOCTL, uintptr(loopFd), CmdSetStatus64, uintptr(unsafe.Pointer(loop.Info))); err != 0 {
-			if err == syscall.EAGAIN && i < maxRetries-1 {
-				// with changes introduces in https://github.com/torvalds/linux/commit/5db470e229e22b7eda6e23b5566e532c96fb5bc3
-				// loop_set_status() can temporarily fail with EAGAIN -> sleep and try again
-				// (cf. https://github.com/karelzak/util-linux/blob/dab1303287b7ebe30b57ccc78591070dad0a85ea/lib/loopdev.c#L1355)
-				time.Sleep(250 * time.Millisecond)
+			if err == syscall.EAGAIN {
+				eagainFound = true
 				continue
 			}
 			// clear associated file descriptor to release the loop device,
 			// best-effort here without error checking because we need the
 			// error from previous ioctl call
 			syscall.Syscall(syscall.SYS_IOCTL, uintptr(loopFd), CmdClrFd, 0)
 			return fmt.Errorf("failed to set loop flags on loop device: %s", syscall.Errno(err))
```

Comment:
In this code I believe we are no longer calling the CLR_FD ioctl  (to release the device) in the situation where we have been issued with an EAGAIN in response to the attempt to set status. This means we still hold the loop device after exiting this block, and AFAICT we could receive EAGAIN from any number of devices, hold on to all of them, and never release any of them (DoSing ourselves and potentially the whole system out of loop devices).

Reply:
I suspect we would be better off restoring (some) of the per-device EGAIN retry behaviour within the `attachFromFile` function (in addition to wrapping the whole attempt to find and use a loop device in retry behaviour as you have done in `AttachFromFile`). 

The fundamental issue with the original code was that it returns on L191 rather than continue-ing on to the next loop device (after perhaps logging a warning). Given that we are willing to find and use any available loop device that works for us, there may be no need to return on any errors at all. OTOH, if we get non-transient errors from _all_ of the loop devices we try, it would probably be best to not bother retrying any of them. Perhaps it is worth re-casting `egainFound` as `transientErrorFound` and checking for both EBUSY and EAGAIN (as in the losetup code: https://github.com/util-linux/util-linux/blob/2401078c9cf9ddec105e44af074cf2fcbd4ef153/sys-utils/losetup.c#L576-L579).